Added Cognito as option for OAuth

bakrale opened the pull request:
This adds Amazon Cognito as an identity provider for OAuth2. I look forward to any feedback you may have.

pudo commented on aleph/oauth.py:
```diff
@@ -7,9 +7,14 @@
 from servicelayer.extensions import get_entry_point
+from jose import jwt as jwt_jose
+from jose import jwk
+from jose.utils import base64url_decode
```
I think we already have two implementations of JWT in use, I'd like to avoid a third. Can you check out https://docs.authlib.org/en/latest/jose/index.html instead? I'm pretty sure authlib also has solutions for some of the key parsing issues you're solving below.

pudo commented on aleph/oauth.py:
```diff
@@ -77,6 +82,65 @@ def handle_google_oauth(provider, oauth_token):
+    #Construct Public Key
+    public_key = jwk.construct(keys[key_index])
+    message, encoded_signature = str(token).rsplit('.', 1)
+    decoded_signature = base64url_decode(encoded_signature.encode('utf-8'))
```
Are you sure there isn't library support for doing this?

pudo commented on aleph/oauth.py:
```diff
@@ -77,6 +82,65 @@ def handle_google_oauth(provider, oauth_token):
+    access_token = get_claims_cognito(oauth_token.get('access_token'),keys)
+
+    groups = set(access_token.get('cognito:groups',[]))
+    user_id = 'cognito:{}'.format(id_token.get('cognito:username'))
```
Can you even get an ID from cogito? We messed this up in Keycloak, but if there's an ID it's strongly recommended to use it.

pudo commented on aleph/oauth.py:
```diff
@@ -77,6 +82,65 @@ def handle_google_oauth(provider, oauth_token):
+        log.warning('Signature verification failed')
+        return False
+
+    claims = jwt_jose.get_unverified_claims(token)
```
Can this use: https://docs.authlib.org/en/latest/jose/jwt.html#jwt-payload-claims-validation ?

bakrale commented:
Thanks for your feedback, I reworked the function to use authlib as suggested, which actually simplified the code a lot! I also added one more config option for specifying the administrator group. I didn't modify keycloak, but that may also be able to take advantage of that settings variable instead of hardcoding 'superuser'. Let me know what you think!

pudo commented:
Yes I really like that additional config option. Wanna quickly throw it into KC and maybe shorten the name from ADMINISTRATOR to ADMIN? Otherwise I'm happy to do it. 

Thanks a lot for the nice re-work, ready to merge.

bakrale commented:
Done!
